### PR TITLE
containerapp: Wait for revision to be ready

### LINKED ISSUE #2398: Validate container app revision after deployment
`azd deploy` should validate the state of a container app during deployment to ensure it is in a `Provisionsed` state and is considered `Healthy`.

Today after the revision is deployed it is assumed to be health.  Azd should perform the following validations:

1. New revision has transitioned into the `Provisioned` state and `Running`
2. New revision is in a `Healthy` state

### PR DESCRIPTION
**Pull request by weikanglim**
For container app deployments, we now wait for revision to be `Running` before proceeding with the rest of the deployment.

To promote interactivity, we log the progress of the revision in terms of replicas running / ready.

Example captured in non-tty mode:
```
Deploying service webfrontend (Updating container app)
Deploying service webfrontend (Waiting for revision to be ready)
Deploying service webfrontend (Waiting for revision to be ready: 0/2 replicas running, 0/2 replicas ready)
Deploying service webfrontend (Waiting for revision to be ready: 1/2 replicas running, 1/2 replicas ready)
Deploying service webfrontend (Fetching endpoints for container app service)
```

When an error occurs, we provide a generic error message to help with troubleshooting:

image

Fixes #2398

### CODE REVIEW

File: cli/azd/pkg/httputil/poll_record.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"os"
+	"time"
+)
+
+// By default, PollDelay is a no-op function that returns the delay as-is.
+// This function is overridden in the record mode to return the delay as 0.
+func PollDelay(delay time.Duration) time.Duration {
+	d := os.Getenv("AZD_TEST_POLL_DELAY")
```
Comment (weikanglim) [inline cli/azd/pkg/httputil/poll_record.go:16]: This deviates from our existing handling of LROs in recording (see `poll_discard.go`). I think we're at a point where it's too expensive to implement response trimmers for these non-standard poll operations, where we have business logic looking for specific fields in the response body.

It does unfortunately mean larger recording files, and more noise within the recorded interactions, but maybe that's okay.

Also open to any ideas here...

**Comment by rajeshkamal5050** [comment]
Nice change!

nit: suggestion change something like,

> You can view console and system logs for the failed revision named 'foo-revision' in Azure Portal: 
>`generated container app specific revision management portal link`

File: cli/azd/pkg/httputil/poll_record.go (hunk 1 above)
Comment (ellismg) [reply cli/azd/pkg/httputil/poll_record.go:16]: Seems like a good approach - I suspect that @hemarina might want to leverage this technique with her AppService changes which are also logically polling an endpoint to check status.

I guess long term we could consider something like the recorder keeping a map of mutators that get to run over the recording snapshot before we save it so they could clean up (or elide) intermediate recordings, but I think this is a fine solution for now.

File: cli/azd/pkg/containerapps/container_app.go [hunk 2]
```diff
@@ -256,6 +285,155 @@ func (cas *containerAppService) AddRevision(
...
+				Err: fmt.Errorf("revision '%s' is in a %s state%s",
+					revisionName,
+					*revision.Properties.RunningState,
+					errSuffix),
+				Suggestion: "To view logs:" +
+					"\n1. Visit " + revisionManagementUrl +
+					"\n2. Click on revision " + fmt.Sprintf("'%s'", revisionName) +
+					"\n3. View console and system logs" +
+					"\nFor more troubleshooting information, visit " +
+					output.WithLinkFormat("https://learn.microsoft.com/en-us/azure/container-apps/troubleshooting"),
```
Comment (ellismg) [inline cli/azd/pkg/containerapps/container_app.go:372]:
```suggestion
					output.WithLinkFormat("https://learn.microsoft.com/azure/container-apps/troubleshooting"),
```

NIT: Remove the en-us

This is a great error message by the way!

File: cli/azd/test/recording/sanitize.go [hunk 3]
```diff
@@ -49,16 +49,17 @@ func sanitizeContainerAppListSecrets(i *cassette.Interaction) error {
...
 		for i := range body.Value {
 			if body.Value[i].Name != nil {
+				sanitizedVal := "SANITIZED"
+
 				if body.Value[i].Value != nil {
 					val := *body.Value[i].Value
 					// Redis requirepass. Sanitize the password, remove other config.
 					if strings.Contains(val, "requirepass ") {
-						body.Value[i].Value = convert.RefOf("requirepass SANITIZED")
+						sanitizedVal = "requirepass SANITIZED"
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/test/recording/sanitize.go:58]: Just checking my understanding - the problem was before the `continue` meant that we skipped over the old L61 and in practice we only sanitized the redis require passvalue, but left all the other values as they were?  Good catch.

**Review by ellismg — APPROVED** [review]
LGTM!  I suspect that @hemarina may find this PR interesting - many of the discussions we were having last week about the AppService status monitoring also seem to be present here (and we came up with similar solutions, like the `func(string)` idea for injecting progress reporters and wiring them back up the deployment status.

Very nice to see how cleanly this all landed, @weikanglim, great work!

File: cli/azd/pkg/containerapps/container_app.go [hunk 4]
```diff
@@ -256,6 +285,155 @@ func (cas *containerAppService) AddRevision(
...
+	if err != nil {
+		return err
+	}
+
+	// Add a policy to capture the raw response. This allows us to get fields (ex: 'properties.runningStateDetails') that are
+	// currently not available in armappcontainers.ContainerAppsRevisionsClientGetRevisionResponse
+	responseCapture := azsdk.ResponseCapturePolicy{}
+
+	revisionsClientOptions := *cas.armClientOptions
+	revisionsClientOptions.PerCallPolicies = append(revisionsClientOptions.PerCallPolicies, &responseCapture)
```
Comment (vhvb1989) [inline cli/azd/pkg/containerapps/container_app.go:308]: I don't think we need a policy for getting the http response. There is a policy already in place for it:
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore/policy#example-WithCaptureResponse

```
// policy.WithCaptureResponse provides a mechanism for obtaining an API's underlying *http.Response
	var respFromCtx *http.Response
	ctx := policy.WithCaptureResponse(context.TODO(), &respFromCtx)
	// make some client method call using the updated context
	// resp, err := client.SomeMethod(ctx, ...)
	// *respFromCtx contains the raw *http.Response returned during the client method call.
	// if the HTTP transport didn't return a response due to an error then *respFromCtx will be nil.
	_ = ctx
```

File: cli/azd/pkg/azsdk/response_capture_policy.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+package azsdk
+
+import (
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+// ResponseCapturePolicy is a policy that captures the http.Response from a request.
+type ResponseCapturePolicy struct {
```
Comment (vhvb1989) [inline cli/azd/pkg/azsdk/response_capture_policy.go:13]: use `policy.WithCaptureResponse` instead

File: cli/azd/pkg/containerapps/container_app.go [hunk 6]
```diff
@@ -256,6 +285,155 @@ func (cas *containerAppService) AddRevision(
...
+			revisionManagementUrl := output.WithLinkFormat(
+				"%s/#@/resource/subscriptions/%s/resourceGroups/%s/providers/Microsoft.App/containerApps/%s/revisionManagement",
+				string(cas.portalUrlBase),
+				subscriptionId,
+				resourceGroupName,
+				appName)
+
+			if v, err := strconv.ParseBool(os.Getenv("AZD_DEMO_MODE")); err == nil && v {
+				revisionManagementUrl = fmt.Sprintf("Revision Management for %s in Azure Portal", appName)
+			}
```
Comment (vhvb1989) [inline cli/azd/pkg/containerapps/container_app.go:360]: oh, I missed this `AZD_DEMO_MODE` strategy.  I whish demo-mode was not checked everywhere but only within the output (like the console).  IMHO, making all components be aware of azd-demo-mode is not cool... seems unrelated to what component's responsibility is. :(

File: cli/azd/pkg/containerapps/container_app.go [hunk 7]
```diff
@@ -256,6 +285,155 @@ func (cas *containerAppService) AddRevision(
...
+			return &internal.ErrorWithSuggestion{
+				Err: fmt.Errorf("revision '%s' is in a %s state%s",
+					revisionName,
+					*revision.Properties.RunningState,
+					errSuffix),
+				Suggestion: "To view logs:" +
+					"\n1. Visit " + revisionManagementUrl +
+					"\n2. Click on revision " + fmt.Sprintf("'%s'", revisionName) +
+					"\n3. View console and system logs" +
+					"\nFor more troubleshooting information, visit " +
```
Comment (vhvb1989) [inline cli/azd/pkg/containerapps/container_app.go:371]: nit.  Consider using `ux.MultilineMessage` or doing something like:
```go
strings.Join([]string{
   "Visit...",
    "Click....",
}, "\n").
```

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/pkg/containerapps/container_app.go (hunk 6 above)
Comment (ellismg) [reply cli/azd/pkg/containerapps/container_app.go:360]: > IMHO, making all components be aware of azd-demo-mode is not cool... seems unrelated to what component's responsibility is. :(

Not opposed to fixing this up over time - putting the checks the few places we needed them vs trying to come up with some generalized console filtering strategy (where perhaps we have a special type for SubscriptionID's and some general thing can see them and filter them) is reasonable for us to go do - I just did the direct thing at the time because there were only a handful of places to audit and it felt like each place needed some level of care to decide "how" to elide the thing.

Perhaps to start on a design there we could just lean into the "I'll give you two options, and demo mode and non demo mode version, and the console decides which to write".

File: cli/azd/pkg/containerapps/container_app.go (hunk 4 above)
Comment (weikanglim) [reply cli/azd/pkg/containerapps/container_app.go:308]: Really really cool! Thanks for the tip and saving us more lines-of-code :)

File: cli/azd/pkg/containerapps/container_app.go (hunk 6 above)
Comment (weikanglim) [reply cli/azd/pkg/containerapps/container_app.go:360]: I agree that over time we should do something different, but I chose the path of "letting the requirements play out in code over time". This avoids bloat and also really forces us to think about the right way to implement the feature. Like @ellismg, I also think having the flexibility to provide the right substitution at the right place that doesn't degrade UX is a really nice thing to have.

One could also hope for macOS iTerm to provide text hyperlink support, and the rest would be history...

File: cli/azd/test/recording/sanitize.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/test/recording/sanitize.go:58]: Indeed, the `continue` was left in the wrong place.. Re-experiencing the haunting of `goto`s...

File: cli/azd/pkg/containerapps/container_app.go (hunk 6 above)
Comment (vhvb1989) [reply cli/azd/pkg/containerapps/container_app.go:360]: > "letting the requirements play out in code over time"

Is that another way to name `tech debt` ? =P

File: cli/azd/pkg/httputil/poll_record.go (hunk 1 above)
Comment (weikanglim) [reply cli/azd/pkg/httputil/poll_record.go:16]: @hemarina's change for deployment status uses an actual LRO / Azure LRO, our existing implementation will automatically handle it which is nice.

The part that gets trickier is for these types of desired-state polling with no unique operation IDs.

> a map of mutators that get to run over the recording snapshot before we save it so they could clean up (or elide) intermediate recordings

This looks pretty close to the existing poll_discard implementation. The thing that was concerning is scaling it to all the different types of desired states. For example, what if two parts of the app was polling for different end states. Without a unique operation ID, it would be hard to track.

As I am writing this comment, I suppose if we wanted to really implement it, we could smuggle a operation ID in record mode and we can automatically trim all except the last response.. Not a bad idea really...

**Comment by weikanglim** [comment]
> Nice change!
> 
> nit: suggestion change something like,
> 
> > You can view console and system logs for the failed revision named 'foo-revision' in Azure Portal:
> > `generated container app specific revision management portal link`

@rajeshkamal5050 That was close to my initial error messaging. One part that gets confusing is if you click on "View details" instead of the "revision name" on the revision management portal, and click the Console / System logs for troubleshooting, it ends up not taking you to container app replica logs.. To avoid that confusion, I wrote the explicit link to click on (which users may still get confused about unfortunately).

File: cli/azd/pkg/containerapps/container_app.go (hunk 7 above)
Comment (weikanglim) [reply cli/azd/pkg/containerapps/container_app.go:368]: Will leave it as-is. `ux.MultilineMessage` wouldn't work here as an inner struct. I do aesthetically prefer string concat here...

File: cli/azd/pkg/containerapps/container_app.go (hunk 6 above)
Comment (weikanglim) [reply cli/azd/pkg/containerapps/container_app.go:357]: In some ways, yes. But I'm not completely sure about the overall impact here... 

Sometimes, waiting for the right moment to implement the right thing with known requirements pays out in long run. But I'll follow up with you to see if anything immediate warrants addressing.

File: cli/azd/pkg/httputil/poll_record.go (hunk 1 above)
Comment (weikanglim) [reply cli/azd/pkg/httputil/poll_record.go]: With the latest change, we call `httputil.PollHeader` which returns a unique polling ID in record mode, and simply `nil` in non-record mode. The responses are then trimmed by the recorder. I think this works out pretty well.
